### DISCUSSION TITLE
Transactions inside transactions, is that possible?

### DISCUSSION THREAD

**Discussion by felinto-dev** [prisma/prisma#12373] [Q&A] [answered]
https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api

I get the following code FROM documentation:

```ts
async function transfer(from: string, to: string, amount: number) {
  return await prisma.$transaction(async (prisma) => {
    // 1. Decrement amount from the sender.
    const sender = await prisma.account.update({
      data: {
        balance: {
          decrement: amount,
        },
      },
      where: {
        email: from,
      },
    })
    // 2. Verify that the sender's balance didn't go below zero.
    if (sender.balance < 0) {
      throw new Error(`${from} doesn't have enough to send ${amount}`)
    }
    // 3. Increment the recipient's balance by amount
    const recipient = prisma.account.update({
      data: {
        balance: {
          increment: amount,
        },
      },
      where: {
        email: to,
      },
    })
    return recipient
  })
}
```

```ts
const [posts, totalPosts] = await prisma.$transaction([
  prisma.post.findMany({ where: { title: { contains: 'prisma' } } }),
  prisma.post.count(),
])
```

Is it possible to do something like it?

```ts
await prisma.$transaction([
    prisma.$transaction(async (prisma) => {
    // 1. Decrement amount from the sender.
    const sender = await prisma.account.update({
      data: {
        balance: {
          decrement: amount,
        },
      },
      where: {
        email: from,
      },
    })
    // 2. Verify that the sender's balance didn't go below zero.
    if (sender.balance < 0) {
      throw new Error(`${from} doesn't have enough to send ${amount}`)
    }
    // 3. Increment the recipient's balance by amount
    const recipient = prisma.account.update({
      data: {
        balance: {
          increment: amount,
        },
      },
      where: {
        email: to,
      },
    })
    return recipient
  })
])
```

I get the following error:

```
Type 'Promise<Model>' is not assignable to type 'PrismaPromise<any>'
```

What do I need it? Why do I expect?

I am running a bulk operation that each insertion needs to use an interactive transaction. ALL insertions MUST FAIL or SUCCESS.

**Comment by nurul3101** [comment]
Hey @felinto-dev 👋 
I don't think that's possible.

Can't you just do all the bulk insertion operations in one single interactive transaction?
Also please note that the interactive transaction is still in preview and not in General Availability(GA).

**Reply by felinto-dev** [reply]
"Can't you just do all the bulk insertion operations in one single interactive transaction?"

Unfortaneley I can't friend. This is the reason because I create this topic :-(

**Comment by janpio** [comment]
What would be the point of wrapping an interactive transaction into a non interactive transaction with just 1 item?

Your code reformatted:
```js
await prisma.$transaction([
    prisma.$transaction(async (prisma) => {
        // 1. Decrement amount from the sender.
        const sender = await prisma.account.update({
            data: {
                balance: {
                    decrement: amount,
                },
            },
            where: {
                email: from,
            },
        })
        // 2. Verify that the sender's balance didn't go below zero.
        if (sender.balance < 0) {
            throw new Error(`${from} doesn't have enough to send ${amount}`)
        }
        // 3. Increment the recipient's balance by amount
        const recipient = prisma.account.update({
            data: {
                balance: {
                    increment: amount,
                },
            },
            where: {
                email: to,
            },
        })
        return recipient
    })
])
```

**Reply by felinto-dev** [reply]
> Your code reformatted

What did you change exactly?

> What would be the point of wrapping an interactive transaction into a non-interactive transaction with just 1 item?

I get this example FROM documentation as I said in the topic :-)
In "real life", I will have more than 1 transaction, of course :-)

**Reply by janpio** [reply]
> What did you change exactly?

Really just intendation - copied to VS Code and then "Format document" so its internal structure is the same.

The documentation does not have 2 wrapped transactions - so the question is _why_ would you want to do that? You can achieve the same inside one interactive transaction really.

**Reply by felinto-dev** [reply]
> The documentation does not have 2 wrapped transactions

Of course. The reason I copy and paste is to avoid confusing anybody else if a specific schema because THIS is not relevant.

I need multiple transactions because I have custom logic for an "upsert" operation.
If the search params NOT exist, I will create a record, but If exists, I just need to do an update IF a specific field is different.

**Reply by janpio** [reply]
And why does that require nested transactions?

**Reply by vladi-stevanovic** [reply]
👋 Hello @felinto-dev  I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! 🙌

**Reply by boredland** [reply]
the reason this kind of nesting is critical to lots of application is, that you might want to encapsulate things in functions, so you'd typically end up with logically nested problems:

```
const criticalHelperFunction = () => {
   return prisma.$transaction(async (prisma) => {
        ...
    })
}

const crititcalFunction = () => {
    return prisma.$transaction(async (prisma) => {
        const crititcalValue = await criticalHelperFunction();
        ...
    })
}
```

To my understanding this is only possible with the current tooling.

**Comment by mtperessoni** [comment, marked as answer]
Hi guys, how u doing?

I have the same cenarious here. I'm devolping a create user service and when this service calls my repository to create the user, i open a transaction. Inside this transaction i need to call my add-profile service to create the user with the respective profiles.

The problem occurs because my repository where i create the user is using the transaction, but when i call my service to create profile it is using the prisma client.

I can't create nasted transactions, but i can overwrite other classes "this" context to use the same transaction as PrismaClient. This is the only way that i found to do what i want. The otherway was duplicating code to use nasted queries how is showed in documentation.

![image](https://user-images.githubusercontent.com/19274336/196526158-d7a754c3-b1c0-4d8f-ae0a-5f07111cd82f.png)
![image](https://user-images.githubusercontent.com/19274336/196526211-35f140c5-89e5-4a35-9079-410b6799f4ab.png)

The $transaction property doesn't exists in Transactions.
My workaround was to use a function where receives a transaction and when $transaction is called i use the Proxy to redirect to our original transaction.

This is the only way that i found that works. When i tryed to use other things like {...tx, $transaction: async (func) => func(tx)} the transaction connection was closed.


Waiting for nasted transactions in prisma

**Reply by pantharshit00** [reply]
Thanks for documenting this. I am going to mark this as a solution for now. 

Original issue related to nested transactions: https://github.com/prisma/prisma/issues/15212

**Reply by avin-kavish** [reply]
This doesn't actually trigger a nested SQL transaction does it? Like,

```sql
BEGIN
-- do stuff
SAVEPOINT xd
-- do stuff
RELEASE SAVEPOINT xd
COMMIT
```

**Comment by Med-H** [comment]
> Hi guys, how u doing?
> 
> 
> 
> I have the same cenarious here. I'm devolping a create user service and when this service calls my repository to create the user, i open a transaction. Inside this transaction i need to call my add-profile service to create the user with the respective profiles.
> 
> 
> 
> The problem occurs because my repository where i create the user is using the transaction, but when i call my service to create profile it is using the prisma client.
> 
> 
> 
> I can't create nasted transactions, but i can overwrite other classes "this" context to use the same transaction as PrismaClient. This is the only way that i found to do what i want. The otherway was duplicating code to use nasted queries how is showed in documentation.
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/19274336/196526158-d7a754c3-b1c0-4d8f-ae0a-5f07111cd82f.png)
> 
> ![image](https://user-images.githubusercontent.com/19274336/196526211-35f140c5-89e5-4a35-9079-410b6799f4ab.png)
> 
> 
> 
> The $transaction property doesn't exists in Transactions.
> 
> My workaround was to use a function where receives a transaction and when $transaction is called i use the Proxy to redirect to our original transaction.
> 
> 
> 
> This is the only way that i found that works. When i tryed to use other things like {...tx, $transaction: async (func) => func(tx)} the transaction connection was closed.
> 
> 
> 
> 
> 
> Waiting for nasted transactions in prisma

Could you please explain more ? I'm also needing something like it in my project, where does the Proxy comes from ?

**Reply by captyoungkyu** [reply]
See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy :)

**Comment by perfect** [comment]
https://github.com/prisma/prisma/issues/17215#issuecomment-1868310016 
This is real work good code.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
